src/rgw/rgw_rest_s3.cc

```diff
@@ -6079,156 +6079,318 @@ bool rgw::auth::s3::S3AnonymousEngine::is_applicable(
-  char * buff = out_string.data();
-
-  if(crc32 ==0) {
-    // the parameters are according to CRC-32 algorithm and its aligned with AWS-cli checksum
-    crc32 = std::unique_ptr<boost::crc_32_type>(new boost::crc_optimal<32, 0x04C11DB7, 0xFFFFFFFF, 0xFFFFFFFF, true, true>);
-  }
 
-  total_byte_len = result_len + 16;//the total is greater in 4 bytes than current size
+  total_byte_len = sql_result.size() + 4; //the total is greater in 4 bytes than current size
 
-  encode_int(&buff[i], total_byte_len, i);//store sizes at the beginning of the buffer
-  encode_int(&buff[i], header_len, i);
+  push_encode_int(total_byte_len, 0);
+  push_encode_int(header_len, 4);
 
   crc32->reset();
-  *crc32 = std::for_each( buff, buff + 8, *crc32 );//crc for starting 8 bytes
+  *crc32 = std::for_each(sql_result.data(), sql_result.data() + 8, *crc32); //crc for starting 8 bytes
   preload_crc = (*crc32)();
-  encode_int(&buff[i], preload_crc, i);
-
-  i += result_len;//advance to the end of payload.
+  push_encode_int(preload_crc, 8);
 
   crc32->reset();
-  *crc32 = std::for_each( buff, buff + i, *crc32 );//crc for payload + checksum
+  *crc32 = std::for_each(sql_result.begin(), sql_result.end(), *crc32); //crc for payload + checksum
   message_crc = (*crc32)();
-  char out_encode[4];
-  encode_int(out_encode, message_crc, i);
-  out_string.append(out_encode,sizeof(out_encode));
 
-  return i;
+  u_int32_t x = htonl(message_crc);
+  sql_result.append(reinterpret_cast<char *>(&x), sizeof(x));
+
+  return sql_result.size();
+}
+
+void aws_response_handler::init_response()
+{ //12 positions for header-crc
+  sql_result = "012345678901";
+}
+
+void aws_response_handler::init_success_response()
+{
+  m_buff_header.clear();
+  header_size = create_header_records();
+  sql_result.append(m_buff_header.c_str(), header_size);
+  sql_result.append(PAYLOAD_LINE);
+}
+
+void aws_response_handler::init_continuation_response()
+{
+  m_buff_header.clear();
+  header_size = create_header_continuation();
+  sql_result.append(m_buff_header, header_size);
+}
+
+void aws_response_handler::init_progress_response()
+{
+  m_buff_header.clear();
+  header_size = create_header_progress();
+  sql_result.append(m_buff_header, header_size);
+}
+
+void aws_response_handler::init_stats_response()
+{
+  m_buff_header.clear();
+  header_size = create_header_stats();
+  sql_result.append(m_buff_header, header_size);
+}
+
+void aws_response_handler::init_end_response()
+{
+  m_buff_header.clear();
+  header_size = create_header_end();
+  sql_result.append(m_buff_header, header_size);
+}
+
+void aws_response_handler::init_error_response(const char *error_message)
+{ //currently not in use. the headers in the case of error, are not extracted by AWS-cli.
+  m_buff_header.clear();
+  header_size = create_error_header_records(error_message);
+  sql_result.append(m_buff_header, header_size);
+}
+
+void aws_response_handler::send_success_response()
+{
+  if (sql_result.size() > strlen(PAYLOAD_LINE))
```

Comment:
i'm not sure this condition is correct (please verify that),
upon init_response and init_success_response,  sql_result.size() will be bigger than payload-line.

the cont-response should send upon no result created by the engine. (and probably the content of sql_result should be replace)